src/atc/eh/EventHubCapture.py

```diff
@@ -0,0 +1,210 @@
+from atc.spark import Spark
+
+from pyspark.sql import functions as f, DataFrame
+
+
+class EventHubCapture(metaclass=SingletonPerId):
+    """Class to access eventhub capture files as table.
+
+    The constructor argument should be given as a key that can be found in
+    the Table Configurator, where the relevant table must be marked with
+    format=avro and with a partitioning from the set of known partitions: y,m,d,h
+    """
+
+    def __init__(self, id: str):
+        tc = TableConfigurator()
+        self.name = tc.table_property(id, "name")
+        self.path = tc.table_property(id, "path")
+        self.format = tc.table_property(id, "format").lower()
+        self.partitioning = tc.table_property(id, "partitioning").lower()
+
+        assert self.format == "avro"
+        self._validate_partitioning()
+
+        if "h" in self.partitioning:
+            self.partition_delta = timedelta(hours=1)
+        else:
+            self.partition_delta = timedelta(days=1)
+
+        self.max_partition = None
+
+    def _validate_partitioning(self, partitioning: str = None):
+        if not partitioning:
+            partitioning = self.partitioning
+
+        if partitioning not in ["ymd", "ymdh"]:
+            raise ValueError("partitioning must be ymd or ymdh")
+
+        return True
+
+    def _create_table(self):
+        """Create the external avro table"""
+        avro_schema = {
+            "namespace": "org.apache.hive",
+            "name": "first_schema",
+            "type": "record",
+            "fields": [
+                {"name": "SequenceNumber", "type": "long"},
+                {"name": "Offset", "type": "string"},
+                {"name": "EnqueuedTimeUtc", "type": "string"},
+                {"name": "Body", "type": "string"},
+            ],
+        }
+        Spark.get().sql(
+            f"""
+            CREATE TABLE {self.name}
+            ({", ".join(f"{c} STRING" for c in self.partitioning)})
+            STORED AS AVRO
+            PARTITIONED BY ({",".join(self.partitioning)})
+            LOCATION "{self.path}"
+            TBLPROPERTIES ('avro.schema.literal'='{json.dumps(avro_schema)}');
+        """
+        )
+
+    def _discover_first_partition(self):
+        """Add the first partition by discovering it from filesystem."""
+        # raise if the table already has partitions
+        if Spark.get().sql(f"SHOW PARTITIONS {self.name}").count():
+            raise AtcEhLogicException("Table partitions already initialized.")
+
+        dbutils = init_dbutils()
+        partition = {}
+        partition_path = ""
+
+        # discover each part
+        for c in self.partitioning:
+            # loop over items in path
+            value = None
+            for file_info in dbutils.fs.ls(self.path + "/" + partition_path):
+                # validate
+                if not file_info.path.startswith("dbfs:"):
+                    raise AtcEhLogicException("Unexpected file paths")
```

Comment:
As it only checks if file_info starts with "dbfs:", which will be the first characters in self.path - could this check be moved up to the __init__ method? 

if not (self.path).startswith("dbfs:"):
                    raise AtcEhLogicException("Unexpected file paths")

Reply:
the rest of the block uses the `.name` part, this is old and I will remove it